### PR TITLE
SolveSATWithGrover: refresh the tasks

### PR DESCRIPTION
**Pull request by tcNickolas**
* Add a task for evaluating one clause of a SAT formula
* Remove 2-SAT oracle task
* Add a section on solving exactly-1 3-SAT problem
* Refactor existing code a bit

### CODE REVIEW

File: SolveSATWithGrover/SolveSATWithGrover.ipynb [hunk 1]
```diff
@@ -342,18 +333,110 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## Part II. Using Grover's algorithm for problems with multiple solutions"
+    "## Part II. Oracles for exactly-1 3-SAT problem\n",
+    "\n",
+    "Exactly-1 3-SAT problem (also known as \"one-in-three 3-SAT\") is a variant of a general 3-SAT problem.\n",
+    "It has a structure similar to 3-SAT problem, but each clause must have *exactly one* true literal \n",
```
Comment (alan-geller) [inline SolveSATWithGrover/SolveSATWithGrover.ipynb]:
```suggestion
    "It has a structure similar to a 3-SAT problem, but each clause must have *exactly one* true literal, \n",
```

File: SolveSATWithGrover/SolveSATWithGrover.ipynb [hunk 2]
```diff
@@ -342,18 +333,110 @@
...
+    "(while in a normal 3-SAT problem each clause must have *at least one* true literal)."
```
Comment (alan-geller) [inline SolveSATWithGrover/SolveSATWithGrover.ipynb]:
```suggestion
    "while in a normal 3-SAT problem each clause must have *at least one* true literal."
```

File: SolveSATWithGrover/Tasks.qs [hunk 3]
```diff
@@ -97,54 +94,44 @@ namespace Quantum.Kata.GroversAlgorithm {
...
-    // and each clause is a disjunction (an OR operation) of two variables or negated variables:
-    //      f(x) = ∧ᵢ (yᵢ₀ ∨ yᵢ₁), yᵢₖ = either xⱼ or ¬xⱼ for some j in {0, ..., N-1}
-    //
-    // For example, XOR function f(x) = x₀ ⊕ x₁ can be represented as a 2-SAT problem 
-    // using N = 2 variables and M = 2 clauses as follows:
-    //      f(x) = x₀ ⊕ x₁ = (x₀ ∨ x₁) ∧ (¬x₀ ∨ ¬x₁)
-    //
+    // Task 1.5. Evaluate one clause of a SAT formula
+    // 
+    // For k-SAT problems, f(x) is represented as a conjunction (an AND operation) of several clauses on N variables,
```
Comment (alan-geller) [inline SolveSATWithGrover/Tasks.qs]:
```suggestion
    // For general SAT problems, f(x) is represented as a conjunction (an AND operation) of several clauses on N variables, 
```

File: SolveSATWithGrover/Tasks.qs [hunk 4]
```diff
@@ -97,54 +94,44 @@ namespace Quantum.Kata.GroversAlgorithm {
...
+    // Example:
+    // The clause clause(x) = x₀ ∨ ¬x₁ can be represented as [(0, true), (1, false)].
+    operation Oracle_SATClause (queryRegister : Qubit[], 
+                                target : Qubit, 
+                                clause : (Int, Bool)[]) : Unit is Adj {        
+        // ...
     }
 
 
     // Task 1.6. k-SAT problem oracle
```
Comment (alan-geller) [inline SolveSATWithGrover/Tasks.qs]: "k-SAT" implies that each clause has at most k variables. What you're building here is really a general SAT problem oracle, allowing an arbitrary number of variables in each clause.
```suggestion
    // Task 1.6. General SAT problem oracle
```

File: SolveSATWithGrover/Tasks.qs [hunk 5]
```diff
@@ -97,54 +94,44 @@ namespace Quantum.Kata.GroversAlgorithm {
...
     //
     // For k-SAT problems, f(x) is represented as a conjunction (an AND operation) of M clauses on N variables,
```
Comment (alan-geller) [inline SolveSATWithGrover/Tasks.qs]:
```suggestion
    // For SAT problems, f(x) is represented as a conjunction (an AND operation) of M clauses on N variables, 
```

File: SolveSATWithGrover/Tasks.qs [hunk 6]
```diff
@@ -164,22 +151,69 @@ namespace Quantum.Kata.GroversAlgorithm {
...
+                          problem : (Int, Bool)[][]) : Unit is Adj {        
+        // ...
+    }
+
+
+    //////////////////////////////////////////////////////////////////
+    // Part II. Oracles for exactly-1 3-SAT problem
+    //////////////////////////////////////////////////////////////////
+    
+    // Exactly-1 3-SAT problem (also known as "one-in-three 3-SAT") is a variant of a general 3-SAT problem.
```
Comment (alan-geller) [inline SolveSATWithGrover/Tasks.qs]:
```suggestion
    // The exactly-1 3-SAT problem (also known as "one-in-three 3-SAT") is a variant of a general 3-SAT problem.
```

File: SolveSATWithGrover/Tasks.qs [hunk 7]
```diff
@@ -164,22 +151,69 @@ namespace Quantum.Kata.GroversAlgorithm {
...
+    // It has a structure similar to 3-SAT problem, but each clause must have exactly one true literal
```
Comment (alan-geller) [inline SolveSATWithGrover/Tasks.qs]:
```suggestion
    // It has a structure similar to a 3-SAT problem, but each clause must have exactly one true literal 
```

File: SolveSATWithGrover/Tasks.qs [hunk 8]
```diff
@@ -164,22 +151,69 @@ namespace Quantum.Kata.GroversAlgorithm {
...
+    // It has a structure similar to 3-SAT problem, but each clause must have exactly one true literal 
+    // (while in a normal 3-SAT problem each clause must have at least one true literal).
+
+
+    // Task 2.1. "Exactly one |1⟩" oracle
+    // Inputs:
+    //      1) 3 qubits in an arbitrary state |x⟩ (input/query register)
+    //      2) a qubit in an arbitrary state |y⟩ (target qubit)
+    //
+    // Goal: Transform state |x, y⟩ into state |x, y ⊕ f(x)⟩ (⊕ is addition modulo 2),
```
Comment (alan-geller) [inline SolveSATWithGrover/Tasks.qs]:
```suggestion
    // Goal: Transform the state |x, y⟩ into the state |x, y ⊕ f(x)⟩ (⊕ is addition modulo 2),
```

File: SolveSATWithGrover/Tasks.qs [hunk 9]
```diff
@@ -164,22 +151,69 @@ namespace Quantum.Kata.GroversAlgorithm {
...
+    // Part III. Using Grover's algorithm for problems with multiple solutions
     //////////////////////////////////////////////////////////////////
     
-    // Task 2.1. Using Grover's algorithm
-    // Goal: Implement Grover's algorithm and use it to find solutions to SAT instances from part I.
+    // Task 3.1. Using Grover's algorithm
+    // Goal: Implement Grover's algorithm and use it to find solutions to SAT instances from parts I and II.
     // This task is not covered by a test and allows you to experiment with running the algorithm.
     //
     // If you want to learn the Grover's algorithm itself, try doing GroversAlgorithm kata first.
```
Comment (alan-geller) [inline SolveSATWithGrover/Tasks.qs]:
```suggestion
    // If you want to learn Grover's algorithm itself, try doing the GroversAlgorithm kata first.
```

File: SolveSATWithGrover/ReferenceImplementation.qs [hunk 10]
```diff
@@ -141,65 +133,129 @@ namespace Quantum.Kata.GroversAlgorithm {
...
-                ApplyWith(ApplyPauliFromBitString(PauliX, true, flip, _), Oracle_Or_Reference(_, ancillaRegister[clauseIndex]), qubits);
-            }
+    // ------------------------------------------------------
+    function GetClauseQubits (queryRegister : Qubit[], clause : (Int, Bool)[]) : (Qubit[], Bool[]) {
+        mutable clauseQubits = new Qubit[Length(clause)];
+        mutable flip = new Bool[Length(clause)];
+        for (varIndex in 0 .. Length(clause) - 1) {
+            let (index, isTrue) = clause[varIndex];
+            // Add the variable used in the clause to the list of variables which we'll need to call the OR oracle
+            let qt = queryRegister[index];
```
[the remark this answers is not in the document]
Comment (tcNickolas) [reply SolveSATWithGrover/ReferenceImplementation.qs]: Good catch!

**Pull request merged**